File: modules/packing/tests/test_repack_uv_charts.cpp
/*
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
#include <lagrange/testing/common.h>

#include <lagrange/IndexedAttribute.h>
#include <lagrange/SurfaceMesh.h>
#include <lagrange/packing/repack_uv_charts.h>
#include <lagrange/views.h>

#include <catch2/matchers/catch_matchers_floating_point.hpp>

#include <array>
#include <cstdint>
#include <limits>
#include <string_view>
#include <vector>

using namespace lagrange;

namespace {

using Scalar = double;
using Index = uint32_t;

template <typename UVScalar = Scalar>
void add_indexed_uv(
    SurfaceMesh<Scalar, Index>& mesh,
    std::vector<UVScalar> uv_values,
    std::vector<Index> uv_indices,
    std::string_view name = "uv")
{
    mesh.template create_attribute<UVScalar>(
        name,
        AttributeElement::Indexed,
        AttributeUsage::UV,
        2,
        {uv_values.data(), uv_values.size()},
        {uv_indices.data(), uv_indices.size()});
}

// Helper to get UV bounding box as {min_u, min_v, max_u, max_v}.
template <typename UVScalar = Scalar>
std::array<UVScalar, 4> get_uv_bbox(SurfaceMesh<Scalar, Index>& mesh, std::string_view name = "uv")
{
    auto& attr = mesh.template get_indexed_attribute<UVScalar>(mesh.get_attribute_id(name));
    auto values = matrix_view(attr.values());
    auto indices = attr.indices().get_all();
    UVScalar min_u = std::numeric_limits<UVScalar>::max();
    UVScalar min_v = std::numeric_limits<UVScalar>::max();
    UVScalar max_u = std::numeric_limits<UVScalar>::lowest();
    UVScalar max_v = std::numeric_limits<UVScalar>::lowest();
    for (size_t i = 0; i < indices.size(); ++i) {
        auto idx = indices[i];
        min_u = std::min(min_u, values(idx, 0));
        min_v = std::min(min_v, values(idx, 1));
        max_u = std::max(max_u, values(idx, 0));
        max_v = std::max(max_v, values(idx, 1));
    }
    return {min_u, min_v, max_u, max_v};
}

} // namespace

TEST_CASE("repack_uv_charts: empty mesh", "[packing][repack]")
{
    SECTION("No vertices, no facets")
    {
        SurfaceMesh<Scalar, Index> mesh;
        add_indexed_uv(mesh, {}, {});

        packing::repack_uv_charts(mesh);
        CHECK(mesh.get_num_vertices() == 0);
        CHECK(mesh.get_num_facets() == 0);
    }

    SECTION("Vertices but no facets")
    {
        SurfaceMesh<Scalar, Index> mesh;
        mesh.add_vertex({0, 0, 0});
        mesh.add_vertex({1, 0, 0});
        mesh.add_vertex({0, 1, 0});

        add_indexed_uv(mesh, {0, 0, 1, 0, 0, 1}, {});

        packing::repack_uv_charts(mesh);
        CHECK(mesh.get_num_vertices() == 3);
        CHECK(mesh.get_num_facets() == 0);
    }
}

TEST_CASE("repack_uv_charts: single chart", "[packing][repack]")
{
    SECTION("Single triangle normalized to unit square")
    {
        SurfaceMesh<Scalar, Index> mesh;
        mesh.add_vertex({0, 0, 0});
        mesh.add_vertex({1, 0, 0});
        mesh.add_vertex({0, 1, 0});
        mesh.add_triangle(0, 1, 2);

        // UVs span [0,2] x [0,2], should be normalized into ~[0,1] (margin shrinks it slightly)
        add_indexed_uv(mesh, {0, 0, 2, 0, 0, 2}, {0, 1, 2});

        packing::repack_uv_charts(mesh);

        auto bbox = get_uv_bbox(mesh);
        CHECK_THAT(bbox[0], Catch::Matchers::WithinAbs(0.0, 1e-6));
        CHECK_THAT(bbox[1], Catch::Matchers::WithinAbs(0.0, 1e-6));
        // Packing margin (default 1e-3) shrinks the chart slightly
        CHECK(bbox[2] <= 1.0 + 1e-6);
        CHECK(bbox[3] <= 1.0 + 1e-6);
        CHECK(bbox[2] > 0.99);
        CHECK(bbox[3] > 0.99);
    }

    SECTION("Two triangles sharing an edge, single chart")
    {
        //  2---3
        //  |\ |
        //  | \|
        //  0---1
        SurfaceMesh<Scalar, Index> mesh;
        mesh.add_vertex({0, 0, 0});
        mesh.add_vertex({1, 0, 0});
        mesh.add_vertex({0, 1, 0});
        mesh.add_vertex({1, 1, 0});
        mesh.add_triangle(0, 1, 2);
        mesh.add_triangle(1, 3, 2);

        // UVs offset to [10,12] x [10,12]
        add_indexed_uv(mesh, {10, 10, 12, 10, 10, 12, 12, 12}, {0, 1, 2, 1, 3, 2});

        packing::repack_uv_charts(mesh);

        auto bbox = get_uv_bbox(mesh);
        CHECK_THAT(bbox[0], Catch::Matchers::WithinAbs(0.0, 1e-6));
        CHECK_THAT(bbox[1], Catch::Matchers::WithinAbs(0.0, 1e-6));
        CHECK(bbox[2] <= 1.0 + 1e-6);
        CHECK(bbox[3] <= 1.0 + 1e-6);
        CHECK(bbox[2] > 0.99);
        CHECK(bbox[3] > 0.99);
    }
}

TEST_CASE("repack_uv_charts: multiple charts", "[packing][repack]")
{
    SECTION("Two disconnected triangles repacked into unit square")
    {
        SurfaceMesh<Scalar, Index> mesh;
        mesh.add_vertex({0, 0, 0});
        mesh.add_vertex({1, 0, 0});
        mesh.add_vertex({0.5, 1, 0});
        mesh.add_vertex({2, 0, 0});
        mesh.add_vertex({3, 0, 0});
        mesh.add_vertex({2.5, 1, 0});
        mesh.add_triangle(0, 1, 2);
        mesh.add_triangle(3, 4, 5);

        // Two separate UV charts (no shared UV vertices or edges)
        add_indexed_uv(mesh, {0, 0, 1, 0, 0.5, 1, 2, 0, 3, 0, 2.5, 1}, {0, 1, 2, 3, 4, 5});

        packing::repack_uv_charts(mesh);

        auto bbox = get_uv_bbox(mesh);
        // All UVs should fit within [0,1] x [0,1]
        CHECK(bbox[0] >= -1e-6);
        CHECK(bbox[1] >= -1e-6);
        CHECK(bbox[2] <= 1.0 + 1e-6);
        CHECK(bbox[3] <= 1.0 + 1e-6);
    }

    SECTION("With pre-computed chart attribute")
    {
        // Two triangles sharing an edge, but forced into separate charts
        SurfaceMesh<Scalar, Index> mesh;
        mesh.add_vertex({0, 0, 0});
        mesh.add_vertex({1, 0, 0});
        mesh.add_vertex({0, 1, 0});
        mesh.add_vertex({1, 1, 0});
        mesh.add_triangle(0, 1, 2);
        mesh.add_triangle(1, 3, 2);

        add_indexed_uv(mesh, {0, 0, 1, 0, 0, 1, 1, 1}, {0, 1, 2, 1, 3, 2});

        // Force different chart ids
        mesh.template create_attribute<Index>(
            "@chart_id",
            AttributeElement::Facet,
            AttributeUsage::Scalar,
            1,
            std::vector<Index>{0, 1});

        packing::RepackOptions opts;
        opts.chart_attribute_name = "@chart_id";
        packing::repack_uv_charts(mesh, opts);

        auto bbox = get_uv_bbox(mesh);
        CHECK(bbox[0] >= -1e-6);
        CHECK(bbox[1] >= -1e-6);
        CHECK(bbox[2] <= 1.0 + 1e-6);
        CHECK(bbox[3] <= 1.0 + 1e-6);
    }
}

TEST_CASE("repack_uv_charts: normalize option", "[packing][repack]")
{
    SECTION("normalize=false preserves original chart scale")
    {
        SurfaceMesh<Scalar, Index> mesh;
        mesh.add_vertex({0, 0, 0});
        mesh.add_vertex({1, 0, 0});
        mesh.add_vertex({0, 1, 0});
        mesh.add_triangle(0, 1, 2);

        // Chart bbox is 5x5; with normalize=false it must remain that size.
        add_indexed_uv(mesh, {0, 0, 5, 0, 0, 5}, {0, 1, 2});

        packing::RepackOptions opts;
        opts.normalize = false;
        packing::repack_uv_charts(mesh, opts);

        auto bbox = get_uv_bbox(mesh);
        // Min still shifted to origin even when normalization is disabled.
        CHECK_THAT(bbox[0], Catch::Matchers::WithinAbs(0.0, 1e-6));
        CHECK_THAT(bbox[1], Catch::Matchers::WithinAbs(0.0, 1e-6));
        CHECK_THAT(bbox[2] - bbox[0], Catch::Matchers::WithinAbs(5.0, 1e-6));
        CHECK_THAT(bbox[3] - bbox[1], Catch::Matchers::WithinAbs(5.0, 1e-6));
    }

    SECTION("normalize=true (default) rescales to unit box")
    {
        SurfaceMesh<Scalar, Index> mesh;
        mesh.add_vertex({0, 0, 0});
        mesh.add_vertex({1, 0, 0});
        mesh.add_vertex({0, 1, 0});
        mesh.add_triangle(0, 1, 2);

        add_indexed_uv(mesh, {0, 0, 5, 0, 0, 5}, {0, 1, 2});

        packing::RepackOptions opts;
        opts.normalize = true;
        packing::repack_uv_charts(mesh, opts);

        auto bbox = get_uv_bbox(mesh);
        CHECK_THAT(bbox[0], Catch::Matchers::WithinAbs(0.0, 1e-6));
        CHECK_THAT(bbox[1], Catch::Matchers::WithinAbs(0.0, 1e-6));
        CHECK(bbox[2] <= 1.0 + 1e-6);
        CHECK(bbox[3] <= 1.0 + 1e-6);
        CHECK(bbox[2] > 0.99);
        CHECK(bbox[3] > 0.99);
    }
}

TEST_CASE("repack_uv_charts: different UV scalar type", "[packing][repack]")
{
    using UVScalar = float;

    SECTION("Float UVs on double mesh")
    {
        SurfaceMesh<Scalar, Index> mesh;
        mesh.add_vertex({0, 0, 0});
        mesh.add_vertex({1, 0, 0});
        mesh.add_vertex({0, 1, 0});
        mesh.add_triangle(0, 1, 2);

        add_indexed_uv<UVScalar>(mesh, {0.f, 0.f, 2.f, 0.f, 0.f, 2.f}, {0, 1, 2});

        packing::repack_uv_charts(mesh);

        auto bbox = get_uv_bbox<UVScalar>(mesh);
        CHECK_THAT(static_cast<double>(bbox[0]), Catch::Matchers::WithinAbs(0.0, 1e-5));
        CHECK_THAT(static_cast<double>(bbox[1]), Catch::Matchers::WithinAbs(0.0, 1e-5));
        CHECK(bbox[2] <= 1.0f + 1e-5f);
        CHECK(bbox[3] <= 1.0f + 1e-5f);
        CHECK(bbox[2] > 0.99f);
        CHECK(bbox[3] > 0.99f);
    }

    SECTION("Float UVs with two charts")
    {
        SurfaceMesh<Scalar, Index> mesh;
        mesh.add_vertex({0, 0, 0});
        mesh.add_vertex({1, 0, 0});
        mesh.add_vertex({0.5, 1, 0});
        mesh.add_vertex({2, 0, 0});
        mesh.add_vertex({3, 0, 0});
        mesh.add_vertex({2.5, 1, 0});
        mesh.add_triangle(0, 1, 2);
        mesh.add_triangle(3, 4, 5);

        add_indexed_uv<UVScalar>(
            mesh,
            {0.f, 0.f, 1.f, 0.f, 0.5f, 1.f, 2.f, 0.f, 3.f, 0.f, 2.5f, 1.f},
            {0, 1, 2, 3, 4, 5});

        packing::repack_uv_charts(mesh);

        auto bbox = get_uv_bbox<UVScalar>(mesh);
        CHECK(bbox[0] >= -1e-5f);
        CHECK(bbox[1] >= -1e-5f);
        CHECK(bbox[2] <= 1.0f + 1e-5f);
        CHECK(bbox[3] <= 1.0f + 1e-5f);
    }
}

File: modules/packing/python/tests/test_repack_uv_charts.py
#
# Copyright 2026 Adobe. All rights reserved.
# This file is licensed to you under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License. You may obtain a copy
# of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under
# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
# OF ANY KIND, either express or implied. See the License for the specific language
# governing permissions and limitations under the License.
#
import lagrange
import numpy as np
import pytest


@pytest.fixture
def make_mesh_with_indexed_uv():
    """Factory fixture to create a mesh with an indexed UV attribute."""

    def _factory(vertices, facets, uv_values, uv_indices, name="uv"):
        mesh = lagrange.SurfaceMesh()
        mesh.add_vertices(np.array(vertices, dtype=np.float64))
        for f in facets:
            mesh.add_triangle(*f)
        mesh.create_attribute(
            name,
            element=lagrange.AttributeElement.Indexed,
            usage=lagrange.AttributeUsage.UV,
            initial_values=np.array(uv_values, dtype=np.float64),
            initial_indices=np.array(uv_indices, dtype=np.uint32),
        )
        return mesh

    return _factory


def _uv_bbox(mesh, name="uv"):
    """Return (min_u, min_v, max_u, max_v) over referenced UV values."""
    attr = mesh.indexed_attribute(name)
    values = np.asarray(attr.values.data)
    indices = np.asarray(attr.indices.data).flatten()
    used = values[indices]
    return used[:, 0].min(), used[:, 1].min(), used[:, 0].max(), used[:, 1].max()


def _assert_no_overlap(mesh, name="uv"):
    """Assert that the mesh's UV charts do not overlap."""
    result = lagrange.bvh.compute_uv_overlap(mesh, uv_attribute_name=name)
    assert not result.has_overlap


class TestRepackUVCharts:
    def test_single_triangle_normalized_to_unit_square(self, make_mesh_with_indexed_uv):
        # UVs span [0, 2] x [0, 2] -> should be normalized into ~[0, 1].
        mesh = make_mesh_with_indexed_uv(
            vertices=[[0, 0, 0], [1, 0, 0], [0, 1, 0]],
            facets=[[0, 1, 2]],
            uv_values=[[0, 0], [2, 0], [0, 2]],
            uv_indices=[[0, 1, 2]],
        )
        lagrange.packing.repack_uv_charts(mesh)

        min_u, min_v, max_u, max_v = _uv_bbox(mesh)
        assert abs(min_u) < 1e-6
        assert abs(min_v) < 1e-6
        assert 0.99 < max_u <= 1.0 - 1e-6
        assert 0.99 < max_v <= 1.0 - 1e-6

    def test_two_disconnected_charts_fit_unit_square(self, make_mesh_with_indexed_uv):
        mesh = make_mesh_with_indexed_uv(
            vertices=[
                [0, 0, 0],
                [1, 0, 0],
                [0.5, 1, 0],
                [2, 0, 0],
                [3, 0, 0],
                [2.5, 1, 0],
            ],
            facets=[[0, 1, 2], [3, 4, 5]],
            uv_values=[[0, 0], [1, 0], [0.5, 1], [2, 0], [3, 0], [2.5, 1]],
            uv_indices=[[0, 1, 2], [3, 4, 5]],
        )
        lagrange.packing.repack_uv_charts(mesh)

        min_u, min_v, max_u, max_v = _uv_bbox(mesh)
        assert abs(min_u) < 1e-6
        assert abs(min_v) < 1e-6
        assert max_u <= 1.0 - 1e-6
        assert max_v <= 1.0 - 1e-6

        # Repacked charts should not overlap each other.
        _assert_no_overlap(mesh)

    def test_with_chart_attribute_name(self, make_mesh_with_indexed_uv):
        """Force two triangles into separate charts via a chart_id attribute."""
        mesh = make_mesh_with_indexed_uv(
            vertices=[[0, 0, 0], [1, 0, 0], [0, 1, 0], [1, 1, 0]],
            facets=[[0, 1, 2], [1, 3, 2]],
            uv_values=[[0, 0], [1, 0], [0, 1], [1, 1]],
            uv_indices=[[0, 1, 2], [1, 3, 2]],
        )
        mesh.create_attribute(
            "@chart_id",
            element=lagrange.AttributeElement.Facet,
            usage=lagrange.AttributeUsage.Scalar,
            initial_values=np.array([0, 1], dtype=np.uint32),
        )
        lagrange.packing.repack_uv_charts(mesh, chart_attribute_name="@chart_id")

        min_u, min_v, max_u, max_v = _uv_bbox(mesh)
        assert min_u >= -1e-6
        assert min_v >= -1e-6
        assert max_u <= 1.0 + 1e-6
        assert max_v <= 1.0 + 1e-6

        # The two forced-apart charts should not overlap after repacking.
        _assert_no_overlap(mesh)

    def test_named_uv_attribute(self, make_mesh_with_indexed_uv):
        """Explicit uv_attribute_name argument should route to the right attribute."""
        mesh = make_mesh_with_indexed_uv(
            vertices=[[0, 0, 0], [1, 0, 0], [0, 1, 0]],
            facets=[[0, 1, 2]],
            uv_values=[[10, 10], [12, 10], [10, 12]],
            uv_indices=[[0, 1, 2]],
            name="my_uv",
        )
        lagrange.packing.repack_uv_charts(mesh, uv_attribute_name="my_uv")

        min_u, min_v, max_u, max_v = _uv_bbox(mesh, name="my_uv")
        assert abs(min_u) < 1e-6
        assert abs(min_v) < 1e-6
        assert max_u <= 1.0 + 1e-6
        assert max_v <= 1.0 + 1e-6

    def test_normalize_false_preserves_scale(self, make_mesh_with_indexed_uv):
        """With normalize=False, output should preserve the original chart scale."""
        mesh = make_mesh_with_indexed_uv(
            vertices=[[0, 0, 0], [1, 0, 0], [0, 1, 0]],
            facets=[[0, 1, 2]],
            uv_values=[[0, 0], [5, 0], [0, 5]],
            uv_indices=[[0, 1, 2]],
        )
        lagrange.packing.repack_uv_charts(mesh, normalize=False)

        min_u, min_v, max_u, max_v = _uv_bbox(mesh)
        # Min is still shifted to the origin, and the 5x5 chart extent is preserved.
        assert np.isclose(min_u, 0.0, atol=1e-6)
        assert np.isclose(min_v, 0.0, atol=1e-6)
        assert np.isclose(max_u - min_u, 5.0, atol=1e-6)
        assert np.isclose(max_v - min_v, 5.0, atol=1e-6)

    def test_margin_shrinks_packed_region(self, make_mesh_with_indexed_uv):
        """Larger margin should leave more whitespace around the packed chart."""
        mesh = make_mesh_with_indexed_uv(
            vertices=[[0, 0, 0], [1, 0, 0], [0, 1, 0]],
            facets=[[0, 1, 2]],
            uv_values=[[0, 0], [1, 0], [0, 1]],
            uv_indices=[[0, 1, 2]],
        )
        lagrange.packing.repack_uv_charts(mesh, margin=0.1)

        _, _, max_u, max_v = _uv_bbox(mesh)
        assert max_u <= 1.0 + 1e-6
        assert max_v <= 1.0 + 1e-6
        # A 0.1 margin should visibly shrink the chart.
        assert max_u < 0.95 or max_v < 0.95

    def test_arguments_after_mesh_are_keyword_only(self, make_mesh_with_indexed_uv):
        """All arguments after ``mesh`` must be passed as keyword arguments."""
        mesh = make_mesh_with_indexed_uv(
            vertices=[[0, 0, 0], [1, 0, 0], [0, 1, 0]],
            facets=[[0, 1, 2]],
            uv_values=[[0, 0], [1, 0], [0, 1]],
            uv_indices=[[0, 1, 2]],
        )
        with pytest.raises(TypeError):
            lagrange.packing.repack_uv_charts(mesh, "uv")  # ty: ignore[too-many-positional-arguments]
